// create-readme/lib/components/pkg.js
// Generated by CoffeeScript 2.3.2
(function() {
  var PkgParser, Promise, logger;

  logger = require('../logger');

  Promise = require('bluebird');

  // Parses package.json information
  module.exports = PkgParser = (function() {
    class PkgParser {
      // Creates a new PkgParser.
      // @param (options) [Object] An optional set of options, currently not used
      constructor(options) {
        this.options = options;
        void 0;
      }

      // Parse package.json data. Selects the data that is important for the template only.
      // @param pkg [Object] package.json data
      // @returns [Promise<Object>] The relevant subset of pkg
      run(pkg) {
        logger.info("Parsing pkg");
        return Promise.resolve({
          name: pkg.name,
          version: pkg.version,
          private: pkg.private,
          preferGlobal: pkg.preferGlobal
        });
      }

    };

    // @property [String] The name of this component
    PkgParser.name = "PkgParser";

    return PkgParser;

  }).call(this);

}).call(this);
